docs/src/proposals/page-indexed-transactions.md

```diff
@@ -0,0 +1,222 @@
+accounts for parallel execution. Therefore the current cap is about 35 accounts
+after accounting for signatures and other transaction metadata.
+
+## Proposed Solution
+
+Introduce a new on-chain account indexing program which stores account address
+mappings and add a new transaction format which supports concise account
+references through on-chain account indexes.
+
+### Account Indexing Program
+
+Here we describe a contract-based solution to the problem, whereby a protocol
+developer or end-user can create collections of related accounts on-chain for
+concise use in a transaction's account inputs. This approach is similar to page
+tables used in operating systems to succinctly map virtual addresses to physical
+memory.
+
+After addresses are stored on-chain in an index account, they may be succinctly
+referenced from a transaction using an index rather than a full 32 byte address.
+This will require a new transaction format to make use of these succinct indexes
+as well as runtime handling for looking up and loading accounts from the
+on-chain indexes.
+
+#### State
+
+Index accounts must be rent-exempt and may not be deleted. Stored addresses
+should be append only so that once an address is stored in an index account, it
+may not be removed later. Index accounts must be pre-allocated since Solana does
+not support reallocation for accounts yet.
+
+Since transactions use a u16 offset to look up addresses, index accounts can
+store up to 2^16 addresses each. Anyone may create an index account of any size
+as long as its big enough to store the necessary metadata. In addition to
+stored addresses, index accounts must also track the latest count of stored
+addresses and an authority which must be a present signer for all index
+modifications.
+
+#### Cost
+
+Since index accounts require caching and special handling in the runtime, they should incur
+higher costs for storage. Cost structure design will be added later.
+
+#### Program controlled indexes
+
+If the authority of an index account is controlled by a program, more
+sophisticated indexes could be built with governance features or price curves
+for new index addresses.
+
+### Versioned Transactions
+
+In order to allow accounts to be referenced more succinctly, the structure of
+serialized transactions must be modified. The new transaction format should not
+affect transaction processing in the Solana VM beyond the increased capacity for
+accounts and instruction invocations. Invoked programs will be unaware of which
+transaction format was used.
+
+The new transaction format must be distinguished from the current transaction
+format.  Current transactions can fit at most 19 signatures (64-bytes each) but
+the message header encodes `num_required_signatures` as a `u8`. Since the upper
+bit of the `u8` will never be set for a valid transaction, we can enable it to
+denote whether a transaction should be decoded with the versioned format or not.
+
+#### New Transaction Format
```

Comment:
A problem I see with the new transaction format is we lose the ability to inspect the ledger and figure the accounts that a transaction requires. With the new format account data is required, ie now `solana-ledger-tool` needs to load a snapshot (or partial snapshot).  This'll have implications on BigTable historical transaction storage as well.

This seems tricky to solve since transactions are recorded into the ledger well before the runtime sees them.  Perhaps we have an epoch-based activation for account indexes, like for stake.  Ideally I think the transactions held by the ledger are fully expanded

Reply:
> A problem I see with the new transaction format is we lose the ability to inspect the ledger and figure the accounts that a transaction requires. With the new format account data is required, ie now solana-ledger-tool needs to load a snapshot (or partial snapshot). This'll have implications on BigTable historical transaction storage as well.

I'm guessing this is an issue when we want to see how many times an account is accessed in a slot or epoch or something? How about we store the addresses accessed through indexes in transaction metadata? I think that would be a nice solution to storing fully expanded transactions as well. The original transaction can be stored as is to preserve signature integrity.

Reply:
> I'm guessing this is an issue when we want to see how many times an account is accessed in a slot or epoch or something?

Any transaction rendering in the explorer will run into this, where what gets deserialized from [getTransaction](https://docs.solana.com/developing/clients/jsonrpc-api#gettransaction) contains account indexes and not the accounts themselves.

Additional transaction metadata, in the case of an indexed transaction, seems like it could work though!  And client changes in the the explorer and CLI to know how to handle decoding the new format

Reply:
Added a brief section about necessary metadata changes here: https://github.com/solana-labs/solana/pull/17103/files#diff-e15f9fcbfd6f181fc2a36cf56d0b5bc1d6837fbd1da1d88a5739d12f6878d549R126

As for clients, I added an RPC section to describe the rough approach we will need to take to transition downstream clients: https://github.com/solana-labs/solana/pull/17103/files#diff-e15f9fcbfd6f181fc2a36cf56d0b5bc1d6837fbd1da1d88a5739d12f6878d549R134

Reply:
thanks